transaction/std/src/signed_contingent_input_builder.rs

```diff
@@ -0,0 +1,2272 @@
+                    MemoType::Unused(_)
+                );
+            }
+
+            // Charlie's MOB output should belong to the correct recipient and have correct
+            // amount and have correct memo
+            {
+                let ss = get_tx_out_shared_secret(
+                    charlie.view_private_key(),
+                    &RistrettoPublic::try_from(&charlie_output.public_key).unwrap(),
+                );
+                let (amount, _) = charlie_output.masked_amount.get_value(&ss).unwrap();
+                assert_eq!(amount, Amount::new(value - Mob::MINIMUM_FEE, Mob::ID));
+
+                let memo = charlie_output.e_memo.unwrap().decrypt(&ss);
+                assert_matches!(
+                    MemoType::try_from(&memo).expect("Couldn't decrypt memo"),
+                    MemoType::Unused(_)
+                );
+            }
+
+            // Charlie's T3 change should belong to the correct recipient and have correct
+            // amount and have correct memo
+            {
+                let ss = get_tx_out_shared_secret(
+                    charlie.view_private_key(),
+                    &RistrettoPublic::try_from(&charlie_change.public_key).unwrap(),
+                );
+                let (amount, _) = charlie_change.masked_amount.get_value(&ss).unwrap();
+                assert_eq!(amount, Amount::new(333, token3));
+
+                let memo = charlie_change.e_memo.unwrap().decrypt(&ss);
+                assert_matches!(
+                    MemoType::try_from(&memo).expect("Couldn't decrypt memo"),
+                    MemoType::Unused(_)
+                );
+            }
+
+            // Alice's T2 output should belong to the correct recipient and have correct
+            // amount and have correct memo
+            {
+                let ss = get_tx_out_shared_secret(
+                    alice.view_private_key(),
+                    &RistrettoPublic::try_from(&alice_output.public_key).unwrap(),
+                );
+                let (amount, _) = alice_output.masked_amount.get_value(&ss).unwrap();
+                assert_eq!(amount, Amount::new(100_000, token2));
+
+                let memo = alice_output.e_memo.unwrap().decrypt(&ss);
+                assert_matches!(
+                    MemoType::try_from(&memo).expect("Couldn't decrypt memo"),
+                    MemoType::Unused(_)
+                );
+            }
+        }
+    }
+
+    #[test]
+    // Test that if you add a signed contingent input, but don't add any of your own
+    // input credentials, it fails with "AllRingsPresigned".
+    // (This is expected because (1) there would be no reason for someone to make
+    // signed inputs like this, and (2) there needs to be at least one
+    // pseudo-output where the transaction builder can choose a blinding factor
+    // for it.)
+    fn test_contingent_input_rules_no_input_credentials_doesnt_work() {
```

Comment:
Theoretically, couldn't this happen naturally by the market maker? E.g. Alice and Bob both make signed contingent inputs that match each other. The market maker just pairs them together and all the rings are presigned and everyone is happy?

Reply:
Maybe we should expect it to work. It might actually be a sensible use case that we could create a transaction using only presigned outputs if they do overlap and in that case the blinding factor seems unnecessary. I guess there is probably a concern about who is paying the fee in that situation.

Reply:
So the only way I can see to permit this is, stop requiring that the sum of blinding factors is zero, instead the Tx contains a scalar which is the sum of blinding factors. Then the balance equation is changed to subtract that. Then this could work.

But it's not clear if it should work, because then we can no longer say that at least one MLSAG signs the whole tx prefix. So in this case the tx would be malleable I think which seems bad.